packages/mobile/src/web3/contracts.ts

```diff
@@ -2,25 +2,61 @@ import { newKitFromWeb3 } from '@celo/contractkit'
 import { privateKeyToAddress } from '@celo/utils/src/address'
 import { Platform } from 'react-native'
 import * as net from 'react-native-tcp'
+import { select, take } from 'redux-saga/effects'
+import sleep from 'sleep-promise'
 import { DEFAULT_FORNO_URL } from 'src/config'
 import { IPC_PATH } from 'src/geth/geth'
-import networkConfig from 'src/geth/networkConfig'
+import { store } from 'src/redux/store'
 import Logger from 'src/utils/Logger'
+import { Actions } from 'src/web3/actions'
+import { contractKitReadySelector, fornoSelector } from 'src/web3/selectors'
 import Web3 from 'web3'
 import { provider } from 'web3-core'
 
-// Logging tag
 const tag = 'web3/contracts'
 
-const web3: Web3 = getWeb3()
-let contractKit = newKitFromWeb3(web3)
+export const web3ForUtils: Web3 = new Web3() // Web3 with no provider
+const contractKitForno = newKitFromWeb3(getWeb3(true))
+const contractKitGeth = newKitFromWeb3(getWeb3(false))
 
-export function getContractKit() {
-  return contractKit
+function getWeb3(fornoMode: boolean): Web3 {
+  Logger.info(
+    `${tag}@getWeb3`,
+    `Initializing web3, platform: ${Platform.OS}, forno mode: ${fornoMode}, provider: ${
+      fornoMode ? DEFAULT_FORNO_URL : 'geth'
+    }`
+  )
+  return fornoMode ? new Web3(getHttpProvider(DEFAULT_FORNO_URL)) : new Web3(getIpcProvider())
+}
+
+export async function getContractKitOutsideGenerator() {
```

Comment:
Needing this feels very hacky, can we find a more elegant solution?

Reply:
Agreed 😃  this would be removed as we finish sagafying contractKit logic. Added a comment explaining its temporary context- I had looked at doing the migration in this PR (moved everything straightforward over) but some will require a bigger refactor. 

The ability to block contractKit calls seems worthwhile beyond just switching providers for forno- for example if there are connectivity issues

Reply:
Yeah Anna and I talked about this a bit, and I think similar to what we found with blocking till geth is ready, it's better to have that ability at the contractkit level too. Theoretically when switching we would want to wait for all "in flight" transactions to complete (waiting for a confirmation, for example) and then swap over